### PR TITLE
Update to ZIO 2.0.0-RC

### LINKED ISSUE #377: Upgrade to ZIO 2.0
ZIO is at Milestone 4, with an RC expected in the next few weeks.
https://github.com/zio/zio/releases/tag/v2.0.0-M4

The API is nearly stable at this point, so any early migration work against this version should pay off towards the official 2.0 release.

The progress is being tracked here:
zio/zio#5470

The Stream Encoding work in progress is the only area where the API might still change before the RC.

To assist with the migration, follow this guide:
https://zio.dev/howto/migrate/zio-2.x-migration-guide/

We are actively working on a ScalaFix rule that will cover the bulk of the simple API changes:
https://github.com/zio/zio/blob/series/2.x/scalafix/rules/src/main/scala/fix/Zio2Upgrade.scala

If you would like assistance with the migration from myself or other ZIO contributors, please let us know!

### PR DESCRIPTION
**Pull request by ches**
Resolves #377

This was largely done with [the Scalafix migration][1] provided by ZIO, with a good number of follow-up fixes that it doesn't cover, in particular removal of the `Blocking` service.

Deprecation warnings elevated by sbt-tpolecat covered several more API changes that the Scalafix didn't.

Since this was mostly mechanical so far, I haven't yet looked closely at where code could optionally be made nicer by updates, such as with new `ZLayer` conveniences. When it comes to new module pattern guidance, zio-kafka appears to already partly follow a "separated interface" style, and as a library I'm not sure how much you'd want to add further source-incompatible changes for users upgrading in order to follow the style immediately (e.g. moving "factory" functions like `Consumer.live` to an exposed case class instead). If you wish to though, I'll be happy to take this further.

~~In draft status until I fix a single test case I've commented out in the second commit, I need to free a bit more time to refactor the test for the removal of `unsafeGet` on `Ref`s (zio/zio#4042).~~

- [x] Fix disabled test
- [x] Check docs (README) for any needed updates
- [ ] Change effect/combinator methods to by-name params? Later PR?

[1]: https://zio.dev/howto/migrate/zio-2.x-migration-guide

### CODE REVIEW

**Comment by adamgfraser** [comment]
@ches Thank you so much for doing this! If you have any more feedback on how the Scalafix migration rule worked for you we can feed that into further improvements there.

**Comment by ches** [comment]
@adamgfraser There were failures on a couple of files in zio-kafka when Scalafix ran (`org.scalameta.invariants.InvariantFailedException`s, which sounds like an internal assertion so not sure if it's a bug of the rule, but I can share the full details someplace appropriate). I'm not sure if that halted things prematurely but it appeared to fail at a file level so nearly all source files processed successfully. A few things I needed to fix:

- Many uses of `Blocking.Service`, in both type and value positions
  - Also `ZLayer.identity[Blocking]` used in layer compositions
  - Some instances of `with Blocking` replaced with `with Any`, which compiles but they can be eliminated entirely
- Imports from `zio.blocking` and uses such as `blocking.effectBlocking` => `ZIO.attemptBlocking`
- The move of `Duration` to base `zio` package, in particular the syntax for `30.seconds`—`DurationSyntax` and `DurationOps` are not declared as `implicit class` in `Duration.scala`, I still didn't find exactly where the enrichments are made available but an `import zio._` works (had to add/change such imports manually)
- Breakage from zio/zio#5537

Overall the Scalafix rule and the detailed migration were a great help!

**Comment by adamgfraser** [comment]
@ches Thank you for the detailed feedback! 🙏

**Comment by ches** [comment]
@iravid Do you have some time to take a look? I'd love to get a milestone published for some PoC of ZIO 2 milestones.

File: src/main/scala/zio/kafka/consumer/Consumer.scala [hunk 1]
```diff
@@ -244,11 +244,11 @@ object Consumer {
...
-    override def consumeWith[R, R1, K, V](
+    override def consumeWith[R: Tag, R1: Tag, K, V](
       subscription: Subscription,
       keyDeserializer: Deserializer[R, K],
       valueDeserializer: Deserializer[R, V],
-      commitRetryPolicy: Schedule[Has[Clock], Any, Any] = Schedule.exponential(1.second) && Schedule.recurs(3)
+      commitRetryPolicy: Schedule[Clock, Any, Any] = Schedule.exponential(1.second) && Schedule.recurs(3)
     )(
       f: (K, V) => URIO[R1, Unit]
     ): ZIO[R with R1, Throwable, Unit] =
```
Comment (ches) [inline src/main/scala/zio/kafka/consumer/Consumer.scala]: I am still dog-at-keyboard with adjusting to the `Has` removal—are the `Tag` context bounds here the best way to support this "given environment" pattern?

Comment (adamgfraser) [reply src/main/scala/zio/kafka/consumer/Consumer.scala]: Yes this looks right to me.

**Comment by svroonland** [comment]
Hi @ches, would you be willing to upgrade to RC2 and resolve the merge conflicts?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -31,7 +31,7 @@ jobs:
       fail-fast: false
       matrix:
         java: ['adopt@1.8', 'adopt@1.11']
-        scala: ['2.12.14', '2.13.6', '3.0.0']
+        scala: ['2.12.15', '2.13.6', '3.1.0']
```
Comment (guizmaii) [inline .github/workflows/ci.yml:34]: Your branch might be out of date with the master branch 🤔

**Comment by svroonland** [comment]
@ches I took the liberty of merging master and upgrading to 2.0.0-RC2. 

Unfortunately there are now some compiler errors I cannot make sense of yet.
https://github.com/zio/zio-kafka/runs/5259607988?check_suite_focus=true appears to be related to the `assertTrue` macro
https://github.com/zio/zio-kafka/runs/5259607958?check_suite_focus=true I'm not sure
https://github.com/zio/zio-kafka/runs/5259607971?check_suite_focus=true related to a `Tag`

**Comment by adamgfraser** [comment]
Opened a PR into this branch to fix the compilation errors.

**Comment by svroonland** [comment]
Thanks @adamgfraser , I cherry-picked onto this PR. There's still the Tag issue left on [scala 3](https://github.com/zio/zio-kafka/runs/5270281999?check_suite_focus=true)  and now some new OOM errors.

**Comment by svroonland** [comment]
Tag issue fixed by adding some [type annotations](https://github.com/zio/zio-kafka/pull/383/commits/75feed2ed140df3f2f239244040a1862e4175e8a). 

A couple of minor failing tests still, will pick those up soon.

**Comment by svroonland** [comment]
I believe some of the tests will be fixed by https://github.com/zio/zio/pull/6396

**Comment by svroonland** [comment]
Alright, a green build!

I talked with Itamar about continuing ZIO 2.0 based development on the `master` branch and creating a `series/0.x` branch for backporting new stuff / bugfixes on the ZIO 1.0-based `zio-kafka` version.

So if there are no objections, I'd like to merge this to master and release it as version `2.0.0-M1` for the early adopters this Saturday (March 5th).

CC @iravid @adamgfraser

**Comment by vigoo** [comment]
Can we make a release with https://github.com/zio/zio-kafka/pull/427 (after an approval of course) before this? Or from a ZIO 1.x maintenance branch? We are putting that in production within days and it would be nice to not do it with a fork.

**Comment by svroonland** [comment]
@vigoo Ah yes of course. I'll approve your PR.

**Comment by adamgfraser** [comment]
@svroonland Sounds great!

**Comment by svroonland** [comment]
@vigoo The "restartStreamsOnRebalancing mode closes all partition streams" test failed after merging master into this branch. I added a `rechunk(1)` in the test (see latest commit) which fixes it. I saw two elements emitted from `partitionedAssignmentStream` when rebalancing.  I suspect that there's a change in stream execution in ZIO 2 that would not execute the `mapZIO` on line 612 and 620 sequentially one by one, but first 612 for both elements and then 620 for both.

Could you check if this is expected behavior for the transactional functionality?

**Comment by vigoo** [comment]
@svroonland yes I think you are right about `mapZIO` and the `rechunk` in the test is correct (even if we change the new `mapZIO` to behave as before)

**Comment by svroonland** [comment]
Alright, then it's time for a release!

**Review by svroonland — APPROVED** [review]

**Comment by ches** [comment]
Thanks all for your help getting this over the line, needless to say now I was out of action for awhile 🤝

**Pull request merged**